opm/grid/common/ZoltanPartition.cpp

```diff
@@ -176,16 +137,239 @@ zoltanGraphPartitionGridOnRoot(const CpGrid& cpgrid,
+                                 &importProcs,    /* Process rank for source of each incoming vertex */
+                                 &importToPart,   /* New partition for each incoming vertex */
+                                 &numExport,      /* Number of vertices I must send to other processes*/
+                                 &exportGlobalGids,  /* Global IDs of the vertices I must send */
+                                 &exportLocalGids,   /* Local IDs of the vertices I must send */
+                                 &exportProcs,    /* Process to which I send each of the vertices */
+                                 &exportToPart);  /* Partition to which each vertex will belong */
+
+        // This is very important: by default, zoltan sets numImport to -1,
+        // as we are only running Zoltan in serial.
+        // In order to make sense of the rest of  the code, this must be set
+        // to 0.
+        numImport = 0;
+
+
+    }
+    MPI_Barrier(cc);
+    std::vector<unsigned int> importGlobalGidsVector;
+    if (cc.rank() == root) {
+        std::vector<int> numberOfExportedVerticesPerProcess(cc.size(), 0);
+        for (int i = 0; i < numExport; ++i) {
+            numberOfExportedVerticesPerProcess[exportToPart[i]]++;
+        }
+        int dummyForRoot = 0;
+        cc.scatter<int>(numberOfExportedVerticesPerProcess.data(), &dummyForRoot, 1, root);
+
+
+        std::vector<int> offsets(cc.size() + 1, 0);
+        std::partial_sum(numberOfExportedVerticesPerProcess.begin(),
+                         numberOfExportedVerticesPerProcess.end(), offsets.begin() + 1);
+        std::vector<unsigned int> globalIndicesToSend(numExport, 0);
+        std::vector<int> currentIndex(cc.size(), 0);
+        for (int i = 0; i < numExport; ++i) {
+            if (exportToPart[i] >= currentIndex.size()) {
+                OPM_THROW(std::runtime_error, "Something wrong with Zoltan decomposition. "
+                          << "Debug information: exportToPart[i] = "
+                          << exportToPart[i] << ", " << "currentIndex.size() = " << currentIndex.size());
+            }
+            const auto index = currentIndex[exportToPart[i]]++ + offsets[exportToPart[i]];
+
+            if (index >= globalIndicesToSend.size()) {
+                OPM_THROW(std::runtime_error, "Something wrong with Zoltan decomposition. "
+                          << "index " << index << ", " << "globalIndicesToSend.size() = "
+                          << globalIndicesToSend.size()
+                          << "\n\noffsets[exportToPart[i]] = " << offsets[exportToPart[i]]
+                          << "\n\ncurrentIndex[exportToPart[i]] = "<< currentIndex[exportToPart[i]]
+                          <<"\n\nexportToPart[i] = " << exportToPart[i]);
+            }
+
+            globalIndicesToSend[index] = exportGlobalGids[i];
+        }
+        std::vector<unsigned int> dummyIndicesForRoot(numExport, 0);
```

Comment:
Isn't this a bit big, if root is receiving 0 data items?

Reply:
True, this can probably be made of size 1.